# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# Define csv path
path = r'C:\Users\lucav\Desktop\Instacart Basket Analysis'

In [3]:
# Import data from orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Import data from products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import data from departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# 03. Data wrangling

In [6]:
# Drop 'eval_set' column
df_ords = df_ords.drop(columns = ['eval_set'])

In [7]:
# Rename 'order_dow' column to 'orders_day_of_week'
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [8]:
# Change 'order_id' type to string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [9]:
# Transpose df_dep
df_dep_t = df_dep.T

In [10]:
# Add index to df_dep
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [11]:
# Create a new header
new_header = df_dep_t.iloc[0]

In [12]:
# Create a new dataframe that only copies over rows beyond the first row
df_dep_t_new = df_dep_t[1:]

In [13]:
# Add the new header to the new dataframe
df_dep_t_new.columns = new_header

In [14]:
# Create a dictionary from 'df_dep_t_new'
data_dict = df_dep_t_new.to_dict('index')

# 04. Task

## I. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [15]:
# Return the data type of each column in df_ords
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [16]:
# Change 'user_id' type to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [17]:
# Verify data 'user_id' type has been changed to string
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

## II. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [18]:
# Print the first 5 rows from df_ords
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [19]:
# Rename 'orders_day_of_week' column to 'day_of_week'
df_ords_prepared = df_ords.rename(columns = {'orders_day_of_week' : 'day_of_week'}, inplace = False)

# Rename 'order_hour_of_day' column to 'hour_of_day'
df_ords_prepared.rename(columns = {'order_hour_of_day' : 'hour_of_day'}, inplace = True)

## III. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [20]:
# Frequency count for 'hour_of_day'
df_ords_prepared['hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: hour_of_day, dtype: int64

**The busiest hour is 10 am, with 288418 placed orders**

## IV. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [21]:
# Print the name corresponding to department_id 4
print(data_dict.get('4'))

{'department': 'produce'}


**Department_id 4 corresponds to 'produce'**

## V. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [22]:
# Identify the department_id corresponding to 'breakfast'
breakfast_id = int(list(data_dict.keys())[list(data_dict.values()).index({'department': 'breakfast'})])

# Create a subset containing only information on breakfast products
df_breakfast = df_prods.loc[df_prods['department_id'].isin([breakfast_id])]

In [23]:
# Print the first 5 rows from df_breakfast
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## VI. They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [24]:
# Identify the department_id corresponding to 'alcohol'
alcohol_id = int(list(data_dict.keys())[list(data_dict.values()).index({'department': 'alcohol'})])

# Identify the department_id corresponding to 'deli'
deli_id = int(list(data_dict.keys())[list(data_dict.values()).index({'department': 'deli'})])

# Identify the department_id corresponding to 'beverages'
beverages_id = int(list(data_dict.keys())[list(data_dict.values()).index({'department': 'beverages'})])

# Identify the department_id corresponding to 'meat seafood'
meat_seafood_id = int(list(data_dict.keys())[list(data_dict.values()).index({'department': 'meat seafood'})])

In [25]:
# Create a subset containing only information on alcohol, deli, beverages, and meat/seafood products
df_dinner = df_prods.loc[df_prods['department_id'].isin([alcohol_id,deli_id,beverages_id,meat_seafood_id])]

In [26]:
# Print the first 5 rows from df_dinner
df_dinner.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## VII. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [27]:
# Display the number of rows and columns of df_dinner
df_dinner.shape

(7650, 5)

**df_dinner has 7650 rows**

## VIII. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [28]:
# Create a subset containing only information on customer with user_id 1
df_ords_user_id_1 = df_ords_prepared.loc[df_ords_prepared['user_id'].isin(['1'])]

In [29]:
# Display all the information for customer with user_id 1
df_ords_user_id_1

order_id user_id  order_number  day_of_week  hour_of_day  \
0   2539329       1             1            2            8   
1   2398795       1             2            3            7   
2    473747       1             3            3           12   
3   2254736       1             4            4            7   
4    431534       1             5            4           15   
5   3367565       1             6            2            7   
6    550135       1             7            1            9   
7   3108588       1             8            1           14   
8   2295261       1             9            1           16   
9   2550362       1            10            4            8   
10  1187899       1            11            4            8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

## IX. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [30]:
# Descriptive statistics
df_ords_user_id_1.describe()

order_number  day_of_week  hour_of_day  days_since_prior_order
count     11.000000    11.000000    11.000000               10.000000
mean       6.000000     2.636364    10.090909               19.000000
std        3.316625     1.286291     3.477198                9.030811
min        1.000000     1.000000     7.000000                0.000000
25%        3.500000     1.500000     7.500000               14.250000
50%        6.000000     3.000000     8.000000               19.500000
75%        8.500000     4.000000    13.000000               26.250000
max       11.000000     4.000000    16.000000               30.000000

## X. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [31]:
df_ords_prepared.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

## XI. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [32]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))